# Analysis for the process of 5 Quizzes                   CLUSTER (START)

In [2]:
import pandas as pd
import numpy as np
import seaborn as sn
import scipy.stats as st
import statsmodels.stats.multicomp as ml
from scipy.stats import f_oneway
from statsmodels.stats.multicomp import pairwise_tukeyhsd
import scikit_posthocs as sp
import matplotlib
import matplotlib.pyplot as plt 
import scipy.stats as stats
%matplotlib inline

In [3]:
#ΕΙΣΑΓΩΓΗ ΔΕΔΟΜΕΝΩΝ
df = pd.read_excel('MOOC3.xlsx', sheet_name= "GENERAL", skiprows=1)
df.head()
#ΕΠΙΛΟΓΗ ΣΥΓΚΕΚΡΙΜΕΝΩΝ ΣΤΗΛΩΝ (total_grade,Cluster (START),Cluster (AFTER),Subgroup,QUIZ  1,QUIZ  2,QUIZ  3,QUIZ  4,QUIZ  5)
df2=df[['total_grade', 'Cluster (START)', 'Cluster (AFTER)', 'Subgroup', 'QUIZ  1', 'QUIZ  2', 'QUIZ  3', 'QUIZ  4', 'QUIZ  5']]
#ΕΠΙΛΟΓΗ SUBGROUP ΝΑ ΕΙΝΑΙ ΙΣΣΟ ΜΕ 1 
#one= df.Subgroup==1
#df_by_one= df[one]
df2

total_grade  Cluster (START)  Cluster (AFTER)  Subgroup QUIZ  1 QUIZ  2  \
0           2.71              2.0              NaN       1.0      10       -   
1            2.7              2.0              NaN       1.0      10    8.67   
2              -              NaN              NaN       NaN       -       -   
3              -              NaN              NaN       NaN       -       -   
4           9.91              2.0              1.0       1.0      10      10   
...          ...              ...              ...       ...     ...     ...   
2655           -              NaN              NaN       NaN       -       -   
2656        1.23              2.0              NaN       1.0       -       -   
2657        8.73              2.0              2.0       1.0      10    9.33   
2658           -              NaN              NaN       NaN       -       -   
2659        1019              NaN              NaN       NaN     NaN     NaN   

     QUIZ  3 QUIZ  4 QUIZ  5  
0          -       -       -  
1          0       0       0  
2          -       -       -  
3          -       -       -  
4         10    8.75      10  
...      ...     ...     ...  
2655       -       -       -  
2656       -       -       -  
2657    7.78      10      10  
2658       -       -       -  
2659     NaN     NaN     NaN  

[2660 rows x 9 columns]

In [4]:
#EΛΕΓΧΟΣ ΓΙΑ ΚΑΘΕ ΜΑΘΗΤΗ ΑΝ ΕΧΕΙ ΑΠΑΝΤΗΣΕΙ ΣΕ ΤΟΥΛΑΧΙΣΤΟΝ 2 QUIZ ΣΥΝΟΛΙΚΑ. ΑΝ ΕΧΕΙ <2 AΠΑΝΤΗΜΕΝΑ QUIZ ΤΟ ΚΑΝΩ DROP 

Quiz_list=["QUIZ  1", "QUIZ  2","QUIZ  3", "QUIZ  4", "QUIZ  5"]


for i in range(df2.shape[0]):
    counter = 0
    if df2.at[i,'Subgroup'] == 1:
        
        for j in Quiz_list:
            
            if df2.at[i,j]!="-":             
                counter+=1                       
        if counter<2:
            
            df2.drop(df.index[i], inplace=True)           
    else:
        df2.drop(df.index[i], inplace=True)
        
          
df2

C:\Users\stell\anaconda3\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


total_grade  Cluster (START)  Cluster (AFTER)  Subgroup QUIZ  1 QUIZ  2  \
1            2.7              2.0              NaN       1.0      10    8.67   
4           9.91              2.0              1.0       1.0      10      10   
10          9.07              3.0              3.0       1.0    8.89      10   
15           3.9              1.0              NaN       1.0    8.89       8   
19          8.67              3.0              1.0       1.0      10       8   
...          ...              ...              ...       ...     ...     ...   
2643        3.11              3.0              NaN       1.0      10    7.33   
2647         9.2              3.0              3.0       1.0      10    9.33   
2648        8.51              3.0              1.0       1.0      10    9.33   
2652        8.89              3.0              3.0       1.0      10      10   
2657        8.73              2.0              2.0       1.0      10    9.33   

     QUIZ  3 QUIZ  4 QUIZ  5  
1          0       0       0  
4         10    8.75      10  
10        10    8.75       6  
15         -       -       -  
19        10    8.75       6  
...      ...     ...     ...  
2643       -       -       -  
2647      10      10      10  
2648    8.06    8.75      10  
2652      10    8.75      10  
2657    7.78      10      10  

[910 rows x 9 columns]

In [5]:
df2quiz=df2[["QUIZ  1", "QUIZ  2","QUIZ  3", "QUIZ  4", "QUIZ  5"]]
df2quiz.replace(to_replace='-',value=float('nan'), inplace=True)
df2quiz['mean'] = df2quiz.mean(axis=1, skipna=True)

df2info=df2[['total_grade', 'Cluster (START)', 'Cluster (AFTER)', 'Subgroup']]

result= [df2info, df2quiz]
dfquiz_total= pd.concat(result, axis=1)

dfquiz_total

C:\Users\stell\anaconda3\lib\site-packages\pandas\core\frame.py:4379: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().replace(
<ipython-input-5-bc718b0420f9>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df2quiz['mean'] = df2quiz.mean(axis=1, skipna=True)


total_grade  Cluster (START)  Cluster (AFTER)  Subgroup  QUIZ  1  \
1            2.7              2.0              NaN       1.0    10.00   
4           9.91              2.0              1.0       1.0    10.00   
10          9.07              3.0              3.0       1.0     8.89   
15           3.9              1.0              NaN       1.0     8.89   
19          8.67              3.0              1.0       1.0    10.00   
...          ...              ...              ...       ...      ...   
2643        3.11              3.0              NaN       1.0    10.00   
2647         9.2              3.0              3.0       1.0    10.00   
2648        8.51              3.0              1.0       1.0    10.00   
2652        8.89              3.0              3.0       1.0    10.00   
2657        8.73              2.0              2.0       1.0    10.00   

      QUIZ  2  QUIZ  3  QUIZ  4  QUIZ  5   mean  
1        8.67     0.00     0.00      0.0  3.734  
4       10.00    10.00     8.75     10.0  9.750  
10      10.00    10.00     8.75      6.0  8.728  
15       8.00      NaN      NaN      NaN  8.445  
19       8.00    10.00     8.75      6.0  8.550  
...       ...      ...      ...      ...    ...  
2643     7.33      NaN      NaN      NaN  8.665  
2647     9.33    10.00    10.00     10.0  9.866  
2648     9.33     8.06     8.75     10.0  9.228  
2652    10.00    10.00     8.75     10.0  9.750  
2657     9.33     7.78    10.00     10.0  9.422  

[910 rows x 10 columns]

In [6]:
#Cluster Start 1 (LOW>4)
cs1=dfquiz_total.groupby('Cluster (START)').get_group(1)

cs1
col=cs1.columns
low=cs1[col[9:10:1]]
low.dropna(inplace=True)

meanlow = low.mean(axis=0)#μεσος ορος βαθμολογίας μαθητων low στa 5 quizzes

low_list=[]
for i, row in low.iterrows():
    low_list.append(low.at[i,'mean'])
low_list

<ipython-input-6-3a70bd9da8b5>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  low.dropna(inplace=True)


[8.445,
 8.445,
 7.306,
 9.422,
 9.2,
 8.3,
 5.138,
 8.844,
 7.95,
 10.0,
 8.32,
 7.0120000000000005,
 10.0,
 7.42,
 9.778,
 8.862,
 7.997999999999999,
 9.734,
 8.034,
 9.068000000000001,
 8.934000000000001,
 7.260000000000001,
 9.068000000000001,
 9.644,
 9.35,
 10.0,
 10.0,
 10.0,
 8.906,
 9.876000000000001,
 9.128,
 9.512,
 8.8325,
 6.99,
 9.866,
 6.772,
 7.394,
 7.590000000000001,
 9.665,
 9.75,
 8.078,
 9.528,
 9.6,
 8.75,
 8.878,
 7.8180000000000005,
 9.866,
 10.0,
 9.778,
 8.89,
 9.778,
 7.63]

In [7]:
#Cluster Start 2 (MEDIUM 4-5)
cs2=dfquiz_total.groupby('Cluster (START)').get_group(2)

cs2
col=cs2.columns
medium=cs2[col[9:10:1]]
medium.dropna(inplace=True)

meanmedium = medium.mean(axis=0)#μεσος ορος βαθμολογίας μαθητων medium στa 5 quizzes

medium_list=[]
for i, row in medium.iterrows():
    medium_list.append(medium.at[i,'mean'])
medium_list

<ipython-input-7-53de837e9fca>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  medium.dropna(inplace=True)


[3.7340000000000004,
 9.75,
 9.556000000000001,
 8.922,
 10.0,
 8.994,
 9.466,
 7.9719999999999995,
 9.734,
 9.876000000000001,
 9.734,
 6.963333333333334,
 8.906,
 8.494,
 9.32,
 8.24,
 10.0,
 10.0,
 10.0,
 8.522,
 9.366,
 9.776666666666666,
 9.366,
 9.75,
 10.0,
 9.234,
 6.115,
 9.75,
 9.216,
 9.866,
 7.78,
 10.0,
 9.23,
 8.228,
 9.498000000000001,
 9.629999999999999,
 8.584,
 10.0,
 9.278,
 8.684000000000001,
 8.966,
 8.434000000000001,
 9.616,
 10.0,
 9.35,
 9.540000000000001,
 8.816666666666668,
 8.478,
 9.084,
 10.0,
 6.944,
 8.212,
 8.494,
 7.853333333333334,
 9.244,
 7.590000000000001,
 5.512,
 8.895,
 9.35,
 9.084,
 8.23,
 10.0,
 9.6,
 9.116,
 10.0,
 9.665,
 8.672,
 8.922,
 9.278,
 10.0,
 7.78,
 9.35,
 9.404,
 9.75,
 9.75,
 9.394,
 9.616,
 9.528,
 9.876000000000001,
 5.885,
 9.75,
 9.466,
 8.594,
 8.878,
 8.7,
 9.809999999999999,
 10.0,
 8.780000000000001,
 9.11,
 8.462,
 8.6025,
 9.406666666666666,
 4.912000000000001,
 8.89,
 7.322,
 9.866,
 10.0,
 9.335,
 10.0,
 8.9340000000

In [8]:
#Cluster Start 3 (LARGE 5-6)
cs3=dfquiz_total.groupby('Cluster (START)').get_group(3)

cs3
col=cs3.columns
large=cs3[col[9:10:1]]
large.dropna(inplace=True)

meanlarge = large.mean(axis=0)#μεσος ορος βαθμολογίας μαθητων large στa 5 quizzes

large_list=[]
for i, row in large.iterrows():
    large_list.append(large.at[i,'mean'])
large_list

<ipython-input-8-46353f9ea41f>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  large.dropna(inplace=True)


[8.728,
 8.55,
 10.0,
 9.616,
 8.534,
 8.462,
 9.734,
 10.0,
 8.95,
 10.0,
 9.306000000000001,
 9.394,
 9.556000000000001,
 8.466,
 10.0,
 9.866,
 6.452,
 8.335,
 9.75,
 9.866,
 9.5,
 8.982,
 10.0,
 10.0,
 7.215999999999999,
 7.109999999999999,
 9.5,
 10.0,
 9.35,
 9.528,
 10.0,
 6.912000000000001,
 9.6,
 8.565999999999999,
 7.234,
 9.278,
 8.212,
 9.6,
 4.706,
 9.306000000000001,
 10.0,
 9.6,
 9.75,
 8.862,
 10.0,
 9.778,
 7.962000000000001,
 9.734,
 7.662000000000001,
 8.356,
 10.0,
 9.638,
 8.89,
 8.39,
 5.999999999999999,
 10.0,
 9.484,
 8.5775,
 7.2780000000000005,
 8.703999999999999,
 10.0,
 9.394,
 10.0,
 9.943999999999999,
 10.0,
 10.0,
 9.11,
 9.012,
 9.6,
 9.422,
 8.95,
 10.0,
 9.75,
 10.0,
 9.334,
 8.734,
 10.0,
 9.6,
 9.778,
 7.0280000000000005,
 8.462,
 9.75,
 10.0,
 10.0,
 9.5,
 8.478,
 10.0,
 9.394,
 8.906,
 9.5,
 9.216,
 6.4,
 9.6,
 10.0,
 5.85,
 10.0,
 9.0,
 8.362,
 9.866,
 10.0,
 10.0,
 9.216,
 8.772,
 7.203333333333333,
 9.866,
 9.112,
 9.022,
 10.0,
 9.35,
 9.778,
 

In [9]:
#Cluster Start 4 (XLARGE <6)
cs4=dfquiz_total.groupby('Cluster (START)').get_group(4)

cs4
col=cs4.columns
xlarge=cs4[col[9:10:1]]
xlarge.dropna(inplace=True)

meanxlarge = xlarge.mean(axis=0)#μεσος ορος βαθμολογίας μαθητων xlarge στa 5 quizzes

xlarge_list=[]
for i, row in xlarge.iterrows():
    xlarge_list.append(xlarge.at[i,'mean'])
xlarge_list

<ipython-input-9-8640ec88437f>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  xlarge.dropna(inplace=True)


[9.528,
 9.734,
 9.366,
 4.67,
 9.012,
 9.276,
 6.546666666666667,
 7.784000000000001,
 10.0,
 10.0,
 10.0,
 9.616,
 9.665,
 6.851999999999999,
 10.0,
 7.67,
 10.0,
 10.0,
 8.094,
 9.6,
 10.0,
 9.6,
 9.6,
 10.0,
 10.0,
 9.445,
 7.7780000000000005,
 9.75,
 9.75,
 9.6,
 10.0,
 10.0,
 10.0,
 9.644,
 9.866,
 10.0,
 8.05,
 9.75,
 10.0,
 10.0,
 8.744,
 10.0,
 9.75,
 4.666666666666667,
 9.012,
 8.906,
 9.866]

# one way ANOVA test

In [10]:
# Shapiro-Wilk normality test
stats.shapiro(low_list), stats.shapiro(medium_list), stats.shapiro(large_list), stats.shapiro(xlarge_list)

#Όλες οι τιμές p εκτός από την ομάδα medium είναι μικρότερες από το όριο a = 0,05, 
#επομένως απορρίπτεται η μηδενική υπόθεση 
#τα δείγματα προέρχονται από πληθυσμούς που δεν ακολουθούν την κανονική κατανομή

(ShapiroResult(statistic=0.9045938849449158, pvalue=0.000528387725353241),
 ShapiroResult(statistic=0.7831756472587585, pvalue=8.716463036431842e-23),
 ShapiroResult(statistic=0.8099809885025024, pvalue=1.5308098048867915e-21),
 ShapiroResult(statistic=0.6710565090179443, pvalue=5.436572081407576e-09))

In [12]:
# Levene variance test  
stats.levene(low_list, medium_list, large_list, xlarge_list)

#Η τιμή p είναι μικρότερη από το όριο a = 0,05, 
#επομένως απορρίπτεται η μηδενική υπόθεση 
#τα δείγματα προέρχονται από πληθυσμούς με διαφορετική διακύμανση

LeveneResult(statistic=0.37861407300445565, pvalue=0.7684499794605844)

In [13]:
#Apply one way ANOVA by calling the f_oneway() method
F, p = stats.f_oneway(low_list, medium_list, large_list, xlarge_list)
print('F statistic = {:5.3f} and probability p = {:5.3f}'.format(F, p)) 

F statistic = 1.690 and probability p = 0.168


# Kruskal Wallis test

In [14]:
stats.kruskal(low_list, medium_list, large_list, xlarge_list)

KruskalResult(statistic=9.528277588706219, pvalue=0.023032429394043742)

# Dunn’s test

In [15]:
#perform Dunn's test using a Bonferonni correction for the p-values

data = [low_list, medium_list, large_list, xlarge_list]
data

sp.posthoc_dunn(data , p_adjust = 'bonferroni')

1         2         3         4
1  1.000000  0.175121  0.068617  0.020104
2  0.175121  1.000000  1.000000  0.484726
3  0.068617  1.000000  1.000000  0.942592
4  0.020104  0.484726  0.942592  1.000000

# Cohen's Effect Size

In [19]:
from numpy import mean
from numpy import var
from math import sqrt

n1,n2= len(low_list), len(medium_list)
s1,s2= var(low_list, ddof=1), var(medium_list, ddof=1)
s= sqrt(((n1-1)*s1+(n2-1)*s2)/(n1+n2-2))
u1,u2= mean(low_list), mean(medium_list)
d= (u1-u2)/s

cohen= abs(d)
if cohen<0.10:
    print('Very Small Effect Size=',cohen)
elif cohen<0.1:
    print('Small Effect Size=',cohen)
elif cohen<0.50:
    print('Medium Effect Size=',cohen)
elif cohen<0.80:
    print('Large Effect Size=',cohen)
else:
    print('Very Large=', cohen)

Medium Effect Size= 0.2299063003757954


In [22]:
n1,n2= len(low_list), len(large_list)
s1,s2= var(low_list, ddof=1), var(large_list, ddof=1)
s= sqrt(((n1-1)*s1+(n2-1)*s2)/(n1+n2-2))
u1,u2= mean(low_list), mean(large_list)
d= (u1-u2)/s

cohen= abs(d)
if cohen<0.10:
    print('Very Small Effect Size=',cohen)
elif cohen<0.1:
    print('Small Effect Size=',cohen)
elif cohen<0.50:
    print('Medium Effect Size=',cohen)
elif cohen<0.80:
    print('Large Effect Size=',cohen)
else:
    print('Very Large=', cohen)

Medium Effect Size= 0.32621858868159176


In [25]:
n1,n2= len(low_list), len(xlarge_list)
s1,s2= var(low_list, ddof=1), var(xlarge_list, ddof=1)
s= sqrt(((n1-1)*s1+(n2-1)*s2)/(n1+n2-2))
u1,u2= mean(low_list), mean(xlarge_list)
d= (u1-u2)/s

cohen= abs(d)
if cohen<0.10:
    print('Very Small Effect Size=',cohen)
elif cohen<0.1:
    print('Small Effect Size=',cohen)
elif cohen<0.50:
    print('Medium Effect Size=',cohen)
elif cohen<0.80:
    print('Large Effect Size=',cohen)
else:
    print('Very Large=', cohen)

Medium Effect Size= 0.33376648464841535


In [29]:
n1,n2= len(medium_list), len(large_list)
s1,s2= var(medium_list, ddof=1), var(large_list, ddof=1)
s= sqrt(((n1-1)*s1+(n2-1)*s2)/(n1+n2-2))
u1,u2= mean(medium_list), mean(large_list)
d= (u1-u2)/s

cohen= abs(d)
if cohen<0.10:
    print('Very Small Effect Size=',cohen)
elif cohen<0.1:
    print('Small Effect Size=',cohen)
elif cohen<0.50:
    print('Medium Effect Size=',cohen)
elif cohen<0.80:
    print('Large Effect Size=',cohen)
else:
    print('Very Large=', cohen)

Very Small Effect Size= 0.06880164777736592


In [32]:
n1,n2= len(medium_list), len(xlarge_list)
s1,s2= var(medium_list, ddof=1), var(xlarge_list, ddof=1)
s= sqrt(((n1-1)*s1+(n2-1)*s2)/(n1+n2-2))
u1,u2= mean(medium_list), mean(xlarge_list)
d= (u1-u2)/s

cohen= abs(d)
if cohen<0.10:
    print('Very Small Effect Size=',cohen)
elif cohen<0.1:
    print('Small Effect Size=',cohen)
elif cohen<0.50:
    print('Medium Effect Size=',cohen)
elif cohen<0.80:
    print('Large Effect Size=',cohen)
else:
    print('Very Large=', cohen)

Medium Effect Size= 0.11649660392833645


In [33]:
n1,n2= len(large_list), len(xlarge_list)
s1,s2= var(large_list, ddof=1), var(xlarge_list, ddof=1)
s= sqrt(((n1-1)*s1+(n2-1)*s2)/(n1+n2-2))
u1,u2= mean(large_list), mean(xlarge_list)
d= (u1-u2)/s

cohen= abs(d)
if cohen<0.10:
    print('Very Small Effect Size=',cohen)
elif cohen<0.1:
    print('Small Effect Size=',cohen)
elif cohen<0.50:
    print('Medium Effect Size=',cohen)
elif cohen<0.80:
    print('Large Effect Size=',cohen)
else:
    print('Very Large=', cohen)

Very Small Effect Size= 0.05738625277145187


# In the first directions of the pairs the results of effect size are <0 , in the opposite directions of the pairs the results of effect size are >0 :
low-medium --> Very Small: -0.2299063003757954             medium-low --> Medium: 0.2299063003757954
low-large --> Very Small: -0.32621858868159176              large-low --> Medium: 0.32621858868159176
low-xlarge --> Very Small: -0.33376648464841535             xlarge-low --> Medium: 0.33376648464841535
medium-large --> Very Small: 0.06880164777736592           large-medium --> Very Small: -0.06880164777736592
medium-xlarge --> Very Small: -0.11649660392833645         xlarge-medium --> Medium: 0.11649660392833645
large-xlarge --> Very Small: -0.05738625277145187           xlarge-large --> Very Small: 0.05738625277145187

# Analysis for the process of 5 exercises                   CLUSTER (AFTER)

In [35]:
#Cluster After 1 (LOW>4)
ca1=dfquiz_total.groupby('Cluster (AFTER)').get_group(1)

ca1
col=ca1.columns
low_after=ca1[col[9:10:1]]
low_after.dropna(inplace=True)

meanlow_after = low_after.mean(axis=0)#μεσος ορος βαθμολογίας μαθητων low στa 5 quizzes

low_after_list=[]
for i, row in low_after.iterrows():
    low_after_list.append(low_after.at[i,'mean'])
low_after_list


<ipython-input-35-e51d5062ee4e>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  low_after.dropna(inplace=True)


[9.75,
 8.55,
 7.306,
 9.876000000000001,
 9.5,
 9.422,
 8.522,
 9.366,
 9.366,
 9.2,
 9.23,
 8.228,
 9.498000000000001,
 10.0,
 8.3,
 8.32,
 9.616,
 7.0120000000000005,
 9.540000000000001,
 8.212,
 4.706,
 9.778,
 10.0,
 8.494,
 8.356,
 10.0,
 8.89,
 7.42,
 8.23,
 9.116,
 8.703999999999999,
 9.394,
 9.778,
 9.876000000000001,
 5.885,
 8.862,
 7.997999999999999,
 9.334,
 8.7,
 10.0,
 8.6025,
 9.778,
 10.0,
 8.934000000000001,
 10.0,
 8.972,
 9.112,
 9.866,
 9.068000000000001,
 9.644,
 5.78,
 9.26,
 9.75,
 9.528,
 9.63,
 9.394,
 10.0,
 10.0,
 9.25,
 8.668000000000001,
 7.618,
 9.35,
 9.528,
 9.306000000000001,
 8.034,
 9.278,
 9.866,
 10.0,
 10.0,
 8.906,
 7.225,
 8.922,
 3.5540000000000007,
 8.15,
 7.853333333333334,
 7.712000000000001,
 9.216,
 8.565999999999999,
 8.644,
 7.99,
 9.128,
 9.616,
 10.0,
 9.35,
 10.0,
 9.75,
 9.216,
 9.059999999999999,
 8.256,
 10.0,
 10.0,
 9.809999999999999,
 8.8325,
 6.99,
 8.744,
 10.0,
 10.0,
 9.262,
 8.928,
 10.0,
 10.0,
 8.55,
 7.928,
 10.0,
 9.778

In [36]:
#Cluster After 2 (MEDIUM 4-5)
ca2=dfquiz_total.groupby('Cluster (AFTER)').get_group(2)

ca2
col=ca2.columns
medium_after=ca1[col[9:10:1]]
medium_after.dropna(inplace=True)

meanmedium_after = medium_after.mean(axis=0)#μεσος ορος βαθμολογίας μαθητων medium στa 5 quizzes

medium_after_list=[]
for i, row in medium_after.iterrows():
    medium_after_list.append(medium_after.at[i,'mean'])
medium_after_list

<ipython-input-36-ff49dcd2942e>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  medium_after.dropna(inplace=True)


[9.75,
 8.55,
 7.306,
 9.876000000000001,
 9.5,
 9.422,
 8.522,
 9.366,
 9.366,
 9.2,
 9.23,
 8.228,
 9.498000000000001,
 10.0,
 8.3,
 8.32,
 9.616,
 7.0120000000000005,
 9.540000000000001,
 8.212,
 4.706,
 9.778,
 10.0,
 8.494,
 8.356,
 10.0,
 8.89,
 7.42,
 8.23,
 9.116,
 8.703999999999999,
 9.394,
 9.778,
 9.876000000000001,
 5.885,
 8.862,
 7.997999999999999,
 9.334,
 8.7,
 10.0,
 8.6025,
 9.778,
 10.0,
 8.934000000000001,
 10.0,
 8.972,
 9.112,
 9.866,
 9.068000000000001,
 9.644,
 5.78,
 9.26,
 9.75,
 9.528,
 9.63,
 9.394,
 10.0,
 10.0,
 9.25,
 8.668000000000001,
 7.618,
 9.35,
 9.528,
 9.306000000000001,
 8.034,
 9.278,
 9.866,
 10.0,
 10.0,
 8.906,
 7.225,
 8.922,
 3.5540000000000007,
 8.15,
 7.853333333333334,
 7.712000000000001,
 9.216,
 8.565999999999999,
 8.644,
 7.99,
 9.128,
 9.616,
 10.0,
 9.35,
 10.0,
 9.75,
 9.216,
 9.059999999999999,
 8.256,
 10.0,
 10.0,
 9.809999999999999,
 8.8325,
 6.99,
 8.744,
 10.0,
 10.0,
 9.262,
 8.928,
 10.0,
 10.0,
 8.55,
 7.928,
 10.0,
 9.778

In [37]:
#Cluster After 3 (LARGE 5-6)
ca3=dfquiz_total.groupby('Cluster (AFTER)').get_group(3)

ca3
col=ca3.columns
large_after=ca3[col[9:10:1]]
large_after.dropna(inplace=True)

meanlarge_after = large_after.mean(axis=0)#μεσος ορος βαθμολογίας μαθητων large στa 5 quizzes

large_after_list=[]
for i, row in large_after.iterrows():
    large_after_list.append(large_after.at[i,'mean'])
large_after_list

<ipython-input-37-ff8457f2363f>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  large_after.dropna(inplace=True)


[8.728,
 9.734,
 10.0,
 8.95,
 10.0,
 9.528,
 9.734,
 9.306000000000001,
 8.466,
 10.0,
 6.452,
 9.734,
 9.866,
 8.982,
 10.0,
 9.5,
 10.0,
 9.6,
 8.565999999999999,
 9.866,
 9.276,
 8.434000000000001,
 7.234,
 8.478,
 7.784000000000001,
 10.0,
 9.75,
 10.0,
 10.0,
 10.0,
 8.862,
 9.734,
 6.944,
 7.662000000000001,
 9.244,
 7.590000000000001,
 5.999999999999999,
 10.0,
 9.616,
 10.0,
 7.2780000000000005,
 10.0,
 10.0,
 9.665,
 10.0,
 9.943999999999999,
 9.404,
 9.75,
 10.0,
 9.394,
 9.012,
 9.6,
 10.0,
 9.466,
 8.734,
 9.75,
 8.478,
 9.394,
 8.906,
 9.5,
 7.322,
 8.193999999999999,
 10.0,
 9.35,
 8.362,
 9.866,
 9.012,
 9.778,
 10.0,
 9.866,
 10.0,
 9.734,
 9.75,
 9.75,
 9.35,
 8.126000000000001,
 10.0,
 9.75,
 9.866,
 8.5,
 9.128,
 10.0,
 8.506,
 9.306000000000001,
 8.922,
 10.0,
 8.094,
 9.734,
 8.906,
 8.416,
 10.0,
 10.0,
 9.6,
 10.0,
 9.040000000000001,
 10.0,
 8.818000000000001,
 7.146000000000001,
 8.45,
 9.866,
 10.0,
 9.6,
 8.906,
 10.0,
 9.616,
 9.6,
 8.212,
 9.866,
 8.744,
 

In [38]:
#Cluster After 4 (XLARGE <6) #ΤΟ ΒΓΑΖΩ ΕΚΤΟΣ ΛΟΓΩ ΜΙΚΡΟΥ ΔΕΙΓΜΑΤΟΣ
ca4=dfquiz_total.groupby('Cluster (AFTER)').get_group(4)

ca4
col=ca4.columns
xlarge_after=ca4[col[9:10:1]]
xlarge_after.dropna(inplace=True)

meanxlarge_after = xlarge_after.mean(axis=0)#μεσος ορος βαθμολογίας μαθητων xlarge στa 5 quizzes

xlarge_after_list=[]
for i, row in xlarge_after.iterrows():
    xlarge_after_list.append(xlarge_after.at[i,'mean'])
xlarge_after_list


<ipython-input-38-1cf7f4f3d55c>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  xlarge_after.dropna(inplace=True)


[8.534,
 9.278,
 6.851999999999999,
 10.0,
 4.912000000000001,
 10.0,
 9.216,
 9.734,
 10.0,
 10.0,
 9.144,
 7.7780000000000005,
 10.0,
 8.728,
 10.0,
 4.666666666666667,
 8.906]

# one way ANOVA test

In [39]:
# Shapiro-Wilk normality test
stats.shapiro(low_after_list), stats.shapiro(medium_after_list), stats.shapiro(large_after_list)

#Όλες οι τιμές p είναι μεγαλύτερες από το όριο a = 0,05, 
#επομένως «αποτυγχάνουμε να απορρίψουμε» τη μηδενική υπόθεση 
#τα δείγματα προέρχονται από πληθυσμούς που ακολουθούν την κανονική κατανομή

(ShapiroResult(statistic=0.8244163990020752, pvalue=3.725215394403264e-13),
 ShapiroResult(statistic=0.8244163990020752, pvalue=3.725215394403264e-13),
 ShapiroResult(statistic=0.7331228852272034, pvalue=8.117533710114869e-20))

In [40]:
# Levene variance test  
stats.levene(low_after_list, medium_after_list, large_after_list)

#Η τιμή p είναι μεγαλύτερη από το όριο a = 0,05, 
#επομένως «αποτυγχάνουμε να απορρίψουμε» τη μηδενική υπόθεση 
#τα δείγματα προέρχονται από πληθυσμούς με την ίδια διακύμανση

LeveneResult(statistic=1.7741707296339808, pvalue=0.17052171812366776)

In [41]:
#Apply one way ANOVA by calling the f_oneway() method
F, p = stats.f_oneway(low_after_list, medium_after_list, large_after_list)
print('F statistic = {:5.3f} and probability p = {:5.3f}'.format(F, p)) 

F statistic = 5.043 and probability p = 0.007


# Kruskal Wallis test

In [42]:
stats.kruskal(low_after_list, medium_after_list, large_after_list)

KruskalResult(statistic=13.353534736514328, pvalue=0.0012598440087564547)

# Dunn’s test

In [43]:
#perform Dunn's test using a Bonferonni correction for the p-values

data = [low_after_list, medium_after_list, large_after_list]
data

sp.posthoc_dunn(data , p_adjust = 'bonferroni')

1         2         3
1  1.000000  1.000000  0.006506
2  1.000000  1.000000  0.006506
3  0.006506  0.006506  1.000000

# Cohen's Effect Size

In [60]:
from numpy import mean
from numpy import var
from math import sqrt

n1,n2= len(low_after_list), len(medium_after_list)
s1,s2= var(low_after_list, ddof=1), var(medium_after_list, ddof=1)
s= sqrt(((n1-1)*s1+(n2-1)*s2)/(n1+n2-2))
u1,u2= mean(low_after_list), mean(medium_after_list)
d= (u1-u2)/s

cohen= abs(d)
if cohen<0.10:
    print('Very Small Effect Size=',cohen)
elif cohen<0.1:
    print('Small Effect Size=',cohen)
elif cohen<0.50:
    print('Medium Effect Size=',cohen)
elif cohen<0.80:
    print('Large Effect Size=',cohen)
else:
    print('Very Large=', cohen)

Very Small Effect Size= 0.0


In [63]:
n1,n2= len(low_after_list), len(large_after_list)
s1,s2= var(low_after_list, ddof=1), var(large_after_list, ddof=1)
s= sqrt(((n1-1)*s1+(n2-1)*s2)/(n1+n2-2))
u1,u2= mean(low_after_list), mean(large_after_list)
d= (u1-u2)/s

cohen= abs(d)
if cohen<0.10:
    print('Very Small Effect Size=',cohen)
elif cohen<0.1:
    print('Small Effect Size=',cohen)
elif cohen<0.50:
    print('Medium Effect Size=',cohen)
elif cohen<0.80:
    print('Large Effect Size=',cohen)
else:
    print('Very Large=', cohen)

Medium Effect Size= 0.26844794363910346


In [67]:
n1,n2= len(medium_after_list), len(large_after_list)
s1,s2= var(medium_after_list, ddof=1), var(large_after_list, ddof=1)
s= sqrt(((n1-1)*s1+(n2-1)*s2)/(n1+n2-2))
u1,u2= mean(medium_after_list), mean(large_after_list)
d= (u1-u2)/s

cohen= abs(d)
if cohen<0.10:
    print('Very Small Effect Size=',cohen)
elif cohen<0.1:
    print('Small Effect Size=',cohen)
elif cohen<0.50:
    print('Medium Effect Size=',cohen)
elif cohen<0.80:
    print('Large Effect Size=',cohen)
else:
    print('Very Large=', cohen)

Medium Effect Size= 0.26844794363910346


# In the first directions of the pairs the results of effect size are <0 , in the opposite directions of the pairs the results of effect size are >0 :
low-medium --> Very Small: 0.0                              medium-low --> Very Small: 0.0
low-large --> Very Small: -0.26844794363910346              large-low --> Medium: 0.26844794363910346
medium-large --> Very Small: -0.26844794363910346           large-medium --> Medium: 0.26844794363910346
